### PR TITLE
Feature/bulk with clerical

### PR DESCRIPTION
**Pull request by iRhysBradbury**

### CODE REVIEW

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/BulkMatchController.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    val optRes = request.body match {
+      case Right(file) => {
+        file.file(multiMatchFormName) map { file =>
+          val lines = Source.fromFile(file.ref.file).getLines()
+          //format
+          //id | address
+          apiClient bulk BulkBody(
+            addresses = lines.map { line =>
+              println(line)
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/BulkMatchController.scala]: remove println

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ClericalToolController.scala [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+/**
+  * Controller class for a single address to be matched
+  * @param conf
+  * @param messagesApi
+  * @param apiClient
+  * @param ec
+  */
+@Singleton
+class ClericalToolController @Inject()(
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ClericalToolController.scala]: indentation

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ClericalToolController.scala [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    )
+  }
+
+  /**
+    * Accept posted form, deal with empty address or pass on to MatchWithInput
+    *
+    * @return result to view or redirect
+    */
+  def doMatch(): Action[AnyContent] = Action.async { implicit request =>
+    val addressText = Option(request.body.asFormUrlEncoded.get("address").mkString).getOrElse("")
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ClericalToolController.scala]: BodyParser Here

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ClericalToolController.scala [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    val optFormat: Option[String] = Try(request.body.asFormUrlEncoded.get("format").mkString).toOption
+    val addressFormat = optFormat.getOrElse("paf")
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ClericalToolController.scala]: do we still have format?

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/views/index.scala.html [hunk 5]
```diff
@@ -8,20 +12,43 @@
...
-            <div class="row">
                 <div class="index-nav">
-                    <article class="box-nav-item">
-                        <a href="@{uk.gov.ons.addressIndex.demoui.controllers.routes.SingleMatchController.showSingleMatchPage()}">
-                            <img style="float:right;" width="280" height="120" src="/assets/images/singlesearch.jpg"/>
-                            <h4><a href="@{uk.gov.ons.addressIndex.demoui.controllers.routes.SingleMatchController.showSingleMatchPage()}">@messages("home.singlehead")</a></h4>
-                            <p>@messages("home.singlebody")</p>
-                        </a>
-                    </article>
+                    @nav.navItem(
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/views/index.scala.html]: Nice!

File: demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/ApplicationHomeTest.scala [hunk 6]
```diff
@@ -9,40 +9,40 @@ import play.api.test.{FakeRequest, WithApplication}
...
-  }
-}
+//
+///**
+//  * Tests for home page
+//  */
+//class ApplicationHomeTest extends PlaySpec with Results {
+//  "Application controller" should {
+//    "include at least one link" in new WithApplication {
+//      // Given
```
Comment (iRhysBradbury) [inline demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/ApplicationHomeTest.scala]: oo. whats happening here.

File: demo-ui/test/uk/gov/ons/addressIndex/demoui/model/ui/NavigationSpec.scala [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+//      )
+//      val MultipleMatchLinkExpected = Link(
+//        href = "/bulkAddresses",
+//        label = "Multiple Match"
+//      )
+//    }
+//  }
+//
+//
+//    "Naviagtion" should {
```
Comment (iRhysBradbury) [inline demo-ui/test/uk/gov/ons/addressIndex/demoui/model/ui/NavigationSpec.scala]: should these be commented out

File: server/test/uk/gov/ons/addressIndex/server/modules/ElasticsearchRepositorySpec2.scala [hunk 8]
```diff
@@ -25,7 +25,7 @@ class ElasticSearchRepositorySpec2 extends FlatSpec with ElasticSugar with Match
   }
 
 
-  it should "produce the correct search definition for a uprn" in {
```
Comment (iRhysBradbury) [inline server/test/uk/gov/ons/addressIndex/server/modules/ElasticsearchRepositorySpec2.scala]: I think this suite can be removed

File: demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/ApplicationHomeTest.scala [hunk 9]
```diff
@@ -9,40 +9,40 @@ import play.api.test.{FakeRequest, WithApplication}
...
-      val response = new ApplicationHomeController(configuration, messagesApi).indexPage().apply(FakeRequest())
-      val content = contentAsString(response)
-
-      // Then
-      status(response) mustBe OK
-      content must include(expectedString)
-    }
-  }
-}
+//
```
Comment (Mironor) [inline demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/ApplicationHomeTest.scala]: May I know what's happening with all those commented tests?

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/BulkMatchController.scala [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+@Singleton
+class BulkMatchController @Inject()(
+  val messagesApi: MessagesApi,
+  conf: DemouiConfigModule,
+  apiClient: AddressIndexClientInstance,
+  classHierarchy: ClassHierarchy
+ )(
+  implicit
+  ec: ExecutionContext,
+  mat: akka.stream.Materializer
```
Comment (Mironor) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/BulkMatchController.scala]: Where's akka used?

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ClericalToolController.scala [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+  }
+
+  /**
+    * Accept posted form, deal with empty address or pass on to MatchWithInput
+    *
+    * @return result to view or redirect
+    */
+  def doMatch(): Action[AnyContent] = Action.async { implicit request =>
+    val optAddress: Option[String] = Try(request.body.asFormUrlEncoded.get("address").mkString).toOption
+    val addressText = optAddress.getOrElse("")
```
Comment (Mironor) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ClericalToolController.scala]: `getOrElse` may be applied directly to the `Try` without going through the `Option`

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ClericalToolController.scala [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    }
+  }
+
+  /**
+    * Perform match by calling API with address string. Can be called directly via get or redirect from form
+    *
+    * @param input
+    * @return result to view
+    */
+  def doMatchWithInput(input: String, formatText: String, page: Option[Int], expand: Option[Int]): Action[AnyContent] = Action.async { implicit request =>
```
Comment (Mironor) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ClericalToolController.scala]: The `formatText` may be removed in my opinion as we fetch for addresses that already contain both of the formats.

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/BulkMatchController.scala (hunk 10 above)
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/BulkMatchController.scala]: implicit Materializer is used on line 52

Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/BulkMatchController.scala]: akka use used by play framework

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ClericalToolController.scala (hunk 12 above)
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ClericalToolController.scala]: looks like a legacy hangover from nag/paf index's

**Pull request merged**